[DA] Simplify runtime predicate collection and extend to all dependence tests

sebpop opened the pull request:
Previously, predicates were collected using a local `Assume` vector. This patch:

1. Removes local `Assume` vector, uses class member `Assumptions` instead.
2. Adds a `getNonRedundantAssumptions()` helper for deduplication.
3. Extends predicate collection to all dependence tests.

fhahn commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3570,6 +3570,37 @@ SCEVUnionPredicate DependenceInfo::getRuntimeAssumptions() const {
+    }
+  }
+
+  return SCEVUnionPredicate(UniqueAssumptions, *SE);
```
Could you add a test that shows the impact of the change? `SCEVUnionPredicate` should already filter out redundant predicates, curious in what cases this doesn't happen

kasuga-fj commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -4036,6 +4049,12 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst,
       return nullptr;
   }
 
+  // If runtime assumptions were added but not allowed, return confused
+  // dependence.
+  if (!UnderRuntimeAssumptions && !Assumptions.empty())
+    return std::make_unique<Dependence>(Src, Dst, getNonRedundantAssumptions());
```
I think using assert like `UnderRuntimeAssumptions || Assumptions.empty()` would be better here. We should not add any runtime assumptions if it's not allowed.

sebpop replied:
> I think using assert like UnderRuntimeAssumptions || Assumptions.empty() would be better here. We should not add any runtime assumptions if it's not allowed.

Correct. I changed the condition to an assert.
This change adds early returns when we know there's need for runtime predicates that are not allowed.

I found out that we were recording runtime predicates also for dependence tests that were failing to prove or disprove any dependence. I modified the code to only record runtime predicates when the dependence test succeeds. There is no point in recording predicates that did not help prove the dependence info.

kasuga-fj commented on llvm/include/llvm/Analysis/DependenceAnalysis.h:
```diff
@@ -355,7 +356,33 @@ class DependenceInfo {
+  /// 4. Maintains the intended cascade behavior of the dependence analysis.
+  SCEVUnionPredicate Assumptions;
+
+  /// Indicates whether runtime assumptions are collected during the analysis.
+  /// When false, dependence tests that would require runtime assumptions fail.
+  bool UnderRuntimeAssumptions;
```
Seems unnecessary.

kasuga-fj commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3584,7 +3584,12 @@ SCEVUnionPredicate DependenceInfo::getRuntimeAssumptions() const {
+  // Set the flag for whether we're allowed to add runtime assumptions.
+  this->UnderRuntimeAssumptions = UnderRuntimeAssumptions;
+
+  // Clear any previous assumptions
+  Assumptions = SCEVUnionPredicate({}, *SE);
```
If it is cleared at the beginning, I think there is no longer any reason to retain it in DependenceInfo

kasuga-fj commented on llvm/test/Analysis/DependenceAnalysis/DifferentOffsets.ll:
```diff
@@ -34,8 +34,6 @@ define i32 @alias_with_parametric_offset(ptr nocapture %A, i64 %n) {
 ; CHECK-NEXT:  Src: %0 = load i32, ptr %A, align 1 --> Dst: %0 = load i32, ptr %A, align 1
 ; CHECK-NEXT:    da analyze - none!
-; CHECK-NEXT:  Runtime Assumptions:
-; CHECK-NEXT:  Equal predicate: (zext i2 (trunc i64 %n to i2) to i64) == 0
```
Maybe the original result was excessive? In any case, I also think it would be better to separate patches.

sebpop replied:
> It's not clear to me why this assumption is dropped?

I have said in another comment above:
> I found out that we were recording runtime predicates also for dependence tests that were failing to prove or disprove any dependence. I modified the code to only record runtime predicates when the dependence test succeeds. There is no point in recording predicates that did not help prove the dependence info.

kasuga-fj reviewed (commented):
High-level thoughts: I think the struct/class that holds these assumptions should be separated from DependenceInfo. Reusing the same assumptions implicitly in each `depends` query may lead to unexpected behavior. Perhaps the implementation of LAA could be a useful reference?

EDIT: To be honest, I'm not entirely sure whether the runtime assumptions feature is promising or not, especially DA doesn't yet have a loop versioning mechanism using collected conditions.

sebpop committed:
[DA] Simplify runtime predicate collection

Removes DependenceInfo::getRuntimeAssumptions(), DependenceInfo::Assumptions,
and the print of "Runtime Assumptions:". The runtime assumptions are still
properly attached to each Dependence result and printed as part of the
per-dependence output.

sjoerdmeijer reviewed (approved):
I believe calculating dependencies with runtime assumption when they are needed is the way forward to handle more cases.
E.g., this is the foundation to interchange galgel from spec, and that's what I want to look in next when we have this available.
So, LGTM.
